src/graphic_model/layouters/lomse_noterests_collisions_fixer.cpp

```diff
@@ -0,0 +1,1149 @@
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+// OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+// SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// For any comment, suggestion or feature request, please contact the manager of
+// the project at cecilios@users.sourceforge.net
+//---------------------------------------------------------------------------------------
+
+#include "lomse_noterests_collisions_fixer.h"
+
+#include "lomse_gm_basic.h"
+#include "lomse_internal_model.h"
+#include "lomse_im_note.h"
+#include "lomse_staffobjs_table.h"
+#include "lomse_shape_note.h"
+#include "lomse_logger.h"
+#include "lomse_score_meter.h"
+
+#include <cmath>   //abs
+using namespace std;
+
+
+namespace lomse
+{
+
+#define LOMSE_NOTERESTS_COLLISIONS_COLOURED     0       //1=dbg, 0=release
+const Color m_colorMoved(0, 192, 0);            //green
+const Color m_colorCollision(0, 0, 255);        //blue
+const Color m_colorWarning(255, 128, 0);        //orange
+const Color m_colorInvestigate(255, 0, 255);    //magenta
+
+
+//=====================================================================================
+//NoterestsCollisionsFixer implementation
+//=====================================================================================
+NoterestsCollisionsFixer::NoterestsCollisionsFixer(GmoShape* pShape,
+                                                   ColStaffObjsEntry* pEntry,
+                                                   ScoreMeter* pMeter)
+    : m_iFirstLine(pEntry->line())
+    , m_pMeter(pMeter)
+    , m_iInstr(pEntry->num_instrument())
+    , m_idxStaff( pMeter->staff_index(m_iInstr, pEntry->staff()) )
+{
+    add_noterest(pShape, pEntry);
+}
+
+//---------------------------------------------------------------------------------------
+NoterestsCollisionsFixer::~NoterestsCollisionsFixer()
+{
+    for (auto note : m_notes)
+        delete note;
+    m_notes.clear();
+}
+
+//---------------------------------------------------------------------------------------
+void NoterestsCollisionsFixer::add_noterest(GmoShape* pShape, ColStaffObjsEntry* pEntry)
+{
+//    LOMSE_ASSERT(pEntry->imo_object()->is_note_rest());
+//    LOMSE_ASSERT(pShape->is_shape_note() || pShape->is_shape_rest());
+
+    ImoNoteRest* pNR = static_cast<ImoNoteRest*>(pEntry->imo_object());
+    int posOnStaff = (pNR->is_note() ? static_cast<GmoShapeNote*>(pShape)->get_pos_on_staff()
```

Comment:
This may be something similar to what I have addressed in #311: if an internal object is a note its shape is not necessarily a `GmoShapeNote`. Perhaps it would be better to check the type for the shape itself (or perhaps also exclude invisible notes from this algorithm)?

Reply:
Tanks. Very good point. I never think about invisible shapes as I still don't assume that there are invisible notes and rests in MusicXML; it is an idea that I do not finish assimilating. I'm contrary to using invisible objects as, to me, it is a trick to solve some deficiency in the model. 

I will review this PR for ensuring invisible notes are excluded.